client/app/dedicatedCloud/datacenter/host/dedicatedCloud-datacenter-host.controller.js

```diff
@@ -1,50 +1,83 @@
+      $q,
+      $scope,
+      $state,
+      $stateParams,
+      constants,
+      DedicatedCloud,
+    ) {
+      this.$q = $q;
+      this.$scope = $scope;
+      this.$state = $state;
+      this.$stateParams = $stateParams;
+      this.constants = constants;
+      this.DedicatedCloud = DedicatedCloud;
+    }
 
-  this.$onInit = () => {
-    this.loading = true;
-    this.constants = constants;
-    return DedicatedCloud.getDatacenterInfoProxy(
-      $stateParams.productId,
-      $stateParams.datacenterId,
-    ).then((datacenter) => {
-      $scope.datacenter.model.commercialRangeName = datacenter.commercialRangeName;
-      $scope.datacenter.model.hasDiscountAMD = DedicatedCloud.hasDiscount($scope.datacenter.model);
-    }).finally(() => {
-      this.loading = false;
-    });
-  };
+    $onInit() {
+      return this.fetchDatacenterInfoProxy();
+    }
 
-  this.orderHost = (datacenter) => {
-    if (constants.target === 'US') {
-      $state.go('app.dedicatedClouds.datacenter.hosts.orderUS');
-    } else {
-      $scope.setAction('datacenter/host/order/dedicatedCloud-datacenter-host-order', datacenter.model, true);
+    fetchDatacenterInfoProxy() {
+      this.loading = true;
+
+      return this.DedicatedCloud
+        .getDatacenterInfoProxy(this.$stateParams.productId, this.$stateParams.datacenterId)
+        .then((datacenter) => {
+          this.$scope.datacenter.model.commercialRangeName = datacenter.commercialRangeName;
+          this.$scope.datacenter.model.hasDiscountAMD = this.DedicatedCloud
+            .hasDiscount(this.$scope.datacenter.model);
+        })
+        .finally(() => {
+          this.loading = false;
+        });
     }
-  };
-});
+
+    loadHosts({ offset, pageSize }) {
+      return this.DedicatedCloud
+        .getPaginatedHosts(
+          this.$stateParams.productId,
+          this.$stateParams.datacenterId,
+          pageSize,
+          offset - 1,
+        )
+        .then((result) => {
+          const hosts = _.get(result, 'list.results');
+
+          return this.$q
+            .all(hosts
+              .map((host) => {
+                if (host.billingType === 'HOURLY') {
+                  return this.DedicatedCloud
+                    .getHostHourlyConsumption(
+                      this.$stateParams.productId,
+                      this.$stateParams.datacenterId,
+                      host.hostId,
+                    )
+                    .then(consumption => _.merge({}, host, consumption))
+                    .catch(() => host);
+                }
+
+                return this.$q.when(host);
+              }))
+            .then(hostsWithConsumption => ({
+              data: hostsWithConsumption,
+              meta: {
+                totalCount: result.count,
+              },
+            }));
+        });
+    }
+
+    orderHost(datacenter) {
+      if (this.constants.target === 'US') {
+        this.$state.go('app.dedicatedClouds.datacenter.hosts.orderUS');
```

Comment:
```suggestion
        return this.$state.go('app.dedicatedClouds.datacenter.hosts.orderUS');
```

Reply:
I won't do it as the issue is there about 20 times and each times it means returning importing `$q` and returning an empty Promise 😞